<a href="https://colab.research.google.com/github/BRIJNANDA1979/Segmenting-Kelp-Forests-/blob/main/Version_2_CNN_for_KELP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Next Step is CNN training with Train 5635 1 band  and rejected 469 7Band data.
469*7 = 3293  rejected  Label 0
5635 Trained   Label 1

Total in training:  8928

In [1]:
import pandas as pd
import os

In [2]:
train_kelp_path = '/content/drive/MyDrive/train_kelp'

In [3]:
train_kelp_files_list = os.listdir(train_kelp_path)

In [4]:
test_satellite_path = '/content/drive/MyDrive/test_satellite'

In [5]:
test_satellite_files_list = os.listdir(test_satellite_path)

In [6]:
Predicted = os.listdir('/content/drive/MyDrive/Predicted_all_bands_files')

In [7]:
len(Predicted)

6699

Rejected files  =   Test_satelleite - Predicted

469 7-Band files

In [8]:
for file in Predicted:
  s = file.split('_')
  ss = s[0] + '_satellite.tif'
  #print(ss)
  if ss in test_satellite_files_list:
    test_satellite_files_list.remove(ss)

In [9]:
len(test_satellite_files_list)

469

In [10]:
rejected_files_list = test_satellite_files_list

In [11]:
rejected_files_list[0:2]

['HP504630_satellite.tif', 'HR677407_satellite.tif']

In [12]:
rejected_list_with_7_bands = []
for i in rejected_files_list:
  for j in range(1,8):
    rejected_list_with_7_bands.append(i.split('.')[0]  + str(j) + '.tif')


In [13]:
rejected_list_with_7_bands[0:8]

['HP504630_satellite1.tif',
 'HP504630_satellite2.tif',
 'HP504630_satellite3.tif',
 'HP504630_satellite4.tif',
 'HP504630_satellite5.tif',
 'HP504630_satellite6.tif',
 'HP504630_satellite7.tif',
 'HR677407_satellite1.tif']

In [14]:
len(train_kelp_files_list)

5635

In [15]:
train_kelp_files_list[0:2]

['VN257517_kelp.tif', 'VN609253_kelp.tif']

In [16]:
X_list = train_kelp_files_list + rejected_list_with_7_bands

In [17]:
len(X_list)

8918

In [18]:
df = pd.DataFrame(X_list, columns=['file_name'])

In [19]:
df['file_name'][5633:5635]

5633    CW256294_kelp.tif
5634    CW399607_kelp.tif
Name: file_name, dtype: object

In [20]:
df['file_name'][5635:5640]

5635    HP504630_satellite1.tif
5636    HP504630_satellite2.tif
5637    HP504630_satellite3.tif
5638    HP504630_satellite4.tif
5639    HP504630_satellite5.tif
Name: file_name, dtype: object

In [21]:
label =[]
for i in range(0,5635):
  label.append(1)
for i in range(5635,8918):
  label.append(0)

In [22]:
df['label'] = label

In [23]:
df.head()

file_name  label
0  VN257517_kelp.tif      1
1  VN609253_kelp.tif      1
2  VN630519_kelp.tif      1
3  VN633188_kelp.tif      1
4  VN645015_kelp.tif      1

In [24]:
df.tail()

file_name  label
8913  HL852722_satellite3.tif      0
8914  HL852722_satellite4.tif      0
8915  HL852722_satellite5.tif      0
8916  HL852722_satellite6.tif      0
8917  HL852722_satellite7.tif      0

In [25]:
df['label'].value_counts()

1    5635
0    3283
Name: label, dtype: int64

In [26]:
pip install rasterio

In [27]:
s = 'GV801444_satellite2.tif'
s = s.split('_')


In [28]:
int(s[1][-5])

2

In [29]:
import rasterio

In [30]:
XX = []
YY = []
for i in df['file_name']:
  #print(i)
  if i in train_kelp_files_list:
    with rasterio.open('/content/drive/MyDrive/train_kelp/' + i) as src:
      print(src.shape)
      print(src.meta)
      print(src.read(1))
      data_array = src.read(1)    ##### _kelp is 1 Band only
      XX.append(data_array)
      YY.append(1)
  else:
    s = i.split('_')
    ss = s[0] + '_satellite.tif'
    band = int(s[1][-5])
    with rasterio.open('/content/drive/MyDrive/test_satellite/' +ss) as src:
        data_array = src.read(band)     #### _satellite is 7 bands 1-7
        XX.append(data_array)
        YY.append(0)


/usr/local/lib/python3.10/dist-packages/rasterio/__init__.py:304: NotGeoreferencedWarning: Dataset has no geotransform, gcps, or rpcs. The identity matrix will be returned.
  dataset = DatasetReader(path, driver=driver, sharing=sharing, **kwargs)


Streaming output truncated to the last 5000 lines.
(350, 350)
{'driver': 'GTiff', 'dtype': 'int8', 'nodata': None, 'width': 350, 'height': 350, 'count': 1, 'crs': None, 'transform': Affine(1.0, 0.0, 0.0,
       0.0, 1.0, 0.0)}
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
(350, 350)
{'driver': 'GTiff', 'dtype': 'int8', 'nodata': None, 'width': 350, 'height': 350, 'count': 1, 'crs': None, 'transform': Affine(1.0, 0.0, 0.0,
       0.0, 1.0, 0.0)}
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
(350, 350)
{'driver': 'GTiff', 'dtype': 'int8', 'nodata': None, 'width': 350, 'height': 350, 'count': 1, 'crs': None, 'transform': Affine(1.0, 0.0, 0.0,
       0.0, 1.0, 0.0)}
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
(350, 350)
{'driver': 'GTiff', 'dtype': 'int8', 'nodata': Non

In [31]:
import numpy as np

In [32]:
xx=  np.asarray(XX)
xx.shape

(8918, 350, 350)

In [33]:
yy = np.asarray(YY)

In [34]:
yy.shape

(8918,)

In [35]:
xx = xx.reshape(xx.shape[0],xx.shape[1],xx.shape[2],1)

In [36]:
xx.shape

(8918, 350, 350, 1)

In [37]:
yy = yy.reshape(yy.shape[0],1)

In [38]:
yy.shape

(8918, 1)

In [39]:
X = xx
y= yy

In [40]:
import keras

In [41]:
model = keras.models.Sequential()

initializers = {

}

model.add(
    keras.layers.Conv2D(
        24, 5, input_shape=(350,350,1),
        activation='relu',
    )
)
model.add( keras.layers.MaxPooling2D(2) )
model.add(
    keras.layers.Conv2D(
        48, 5, activation='relu',
    )
)
model.add( keras.layers.MaxPooling2D(2) )
model.add(
    keras.layers.Conv2D(
        96, 5, activation='relu',
    )
)
model.add( keras.layers.Flatten() )
model.add( keras.layers.Dropout(0.2) )

model.add( keras.layers.Dense(
    1, activation='sigmoid',
    )
)

model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 346, 346, 24)      624       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 173, 173, 24)     0         
 )                                                               
                                                                 
 conv2d_1 (Conv2D)           (None, 169, 169, 48)      28848     
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 84, 84, 48)       0         
 2D)                                                             
                                                                 
 conv2d_2 (Conv2D)           (None, 80, 80, 96)        115296    
                                                                 
 flatten (Flatten)           (None, 614400)            0

In [42]:
from sklearn.model_selection import train_test_split

In [43]:
 X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=33)

In [44]:
X_train.shape

(7134, 350, 350, 1)

In [45]:
y_train.shape

(7134, 1)

In [46]:
X_test.shape

(1784, 350, 350, 1)

In [47]:
y_test.shape

(1784, 1)

In [48]:
import tensorflow.keras.optimizers as optimizers

optimizer = optimizers.Adam(learning_rate=0.001)

In [49]:
model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])

In [50]:
# Fit the model
model.fit(X, y, validation_split=0.33, epochs=1, batch_size=100)

60/60 [==============================] - 385s 6s/step - loss: 447.6172 - accuracy: 0.9655 - val_loss: 7087.2559 - val_accuracy: 0.7642


In [51]:
model.fit(X_train, y_train, validation_data=(X_test,y_test), epochs=1, batch_size=100)

72/72 [==============================] - 449s 6s/step - loss: 36.3731 - accuracy: 0.9372 - val_loss: 0.4164 - val_accuracy: 0.9563


In [52]:
prediction = model.predict(X_train)

223/223 [==============================] - 46s 205ms/step


In [53]:
y_train[5]



array([1])

Using CNN let us verify already Predicted using CVI....


In [54]:
Predicted_list_with_7_bands  = Predicted

In [55]:
PX = []
XP = []    ### Predicted data
YP = []
for i in Predicted_list_with_7_bands:
  PX.append(i)
for i in PX:
  #print(i)
  s = i.split('_')
  ss = s[0] + '_satellite.tif'
  band = int(s[1][-5])
  with rasterio.open('/content/drive/MyDrive/test_satellite/' +ss) as src:
      data_array = src.read(band)     #### _satellite is 7 bands 1-7
      XP.append(data_array)
      YP.append(1)


/usr/local/lib/python3.10/dist-packages/rasterio/__init__.py:304: NotGeoreferencedWarning: Dataset has no geotransform, gcps, or rpcs. The identity matrix will be returned.
  dataset = DatasetReader(path, driver=driver, sharing=sharing, **kwargs)


In [56]:
XP = np.asarray(XP)

In [57]:
XP.shape

(6699, 350, 350)

In [58]:
XP = XP.reshape(XP.shape[0],XP.shape[1],XP.shape[2],1)

In [59]:
XP.shape

(6699, 350, 350, 1)

In [60]:
YP = np.asarray(YP)

In [61]:
YP = YP.reshape(YP.shape[0],1)

In [62]:
YP.shape

(6699, 1)

In [63]:
Prediction_by_CNN = model.predict(XP)


210/210 [==============================] - 44s 207ms/step


In [66]:
len(Prediction_by_CNN)

6699

In [64]:
Prediction_by_CNN[0:10]

array([[0.       ],
       [0.       ],
       [0.       ],
       [0.       ],
       [0.9939565],
       [0.       ],
       [0.       ],
       [0.       ],
       [0.       ],
       [0.       ]], dtype=float32)

In [68]:
count=0
for i in Prediction_by_CNN:
  if i[0] < 0.5 :
    count = count + 1
print(count)

5925


In [69]:
Predicted_list_with_7_bands[0:10]               ##########3   GC205345_satellite1.tif

['VU293076_satellite2.tif',
 'VU293076_satellite3.tif',
 'VU293076_satellite4.tif',
 'VU293076_satellite5.tif',
 'VU293076_satellite6.tif',
 'VU293076_satellite7.tif',
 'WI130359_satellite1.tif',
 'WI130359_satellite2.tif',
 'WI130359_satellite3.tif',
 'WI130359_satellite4.tif']

In [72]:
loc =0
for file in Predicted_list_with_7_bands:
  if(file == 'GC205345_satellite1.tif'):
    print(file)
    print(loc)
    print(Prediction_by_CNN[loc])
    break

  loc = loc + 1

GC205345_satellite1.tif
573
[0.]


In [71]:
loc =0                                            ###########3 AC799254_satellite.tif
for file in Predicted_list_with_7_bands:
  if(file == 'AC799254_satellite.tif'):
    print(file)
    print(loc)
    break
  loc = loc + 1

In [75]:
Df_Predictions = pd.DataFrame()
Df_Predictions['images'] = Predicted_list_with_7_bands
label_list = []
for file in Predicted_list_with_7_bands:
  label_list.append(1)
Df_Predictions['Y_Predicted_by_CVI'] = label_list
label_CNN =[]
for i in Prediction_by_CNN:
  if i[0] < 0.5 :
    label_CNN.append(0)
  else:
    label_CNN.append(1)
Df_Predictions['Y_Predicted_by_CNN'] = label_CNN
Df_Predictions.to_csv('/content/drive/MyDrive/Kelp_Predicted_by_CNN/Predictions_by_CNNversion_2_.csv')

In [77]:
Df_Predictions.describe()

Y_Predicted_by_CVI  Y_Predicted_by_CNN
count              6699.0         6699.000000
mean                  1.0            0.115540
std                   0.0            0.319696
min                   1.0            0.000000
25%                   1.0            0.000000
50%                   1.0            0.000000
75%                   1.0            0.000000
max                   1.0            1.000000

In [76]:
Df_Predictions.head()

images  Y_Predicted_by_CVI  Y_Predicted_by_CNN
0  VU293076_satellite2.tif                   1                   0
1  VU293076_satellite3.tif                   1                   0
2  VU293076_satellite4.tif                   1                   0
3  VU293076_satellite5.tif                   1                   0
4  VU293076_satellite6.tif                   1                   1

After running CNN on predictions 6699 made by CVI,
5925 ----> 1
774 ------>  0

So need to remove 774 from 6699 predicetd

In [78]:
Df_Predictions['Y_Predicted_by_CNN'].value_counts()

0    5925
1     774
Name: Y_Predicted_by_CNN, dtype: int64